### PR TITLE
Memory allocator for tmpfs

### PR DESCRIPTION
**Pull request by xThaid**
There should be a dedicated memory allocator for tmpfs as it currently don't have any.

Work is in the very early stage and for now this PR is to present the idea since it should be discussed.

### CODE REVIEW

File: sys/kern/tmpfs.c [hunk 1]
```diff
@@ -536,22 +673,51 @@ static int tmpfs_get_vnode(mount_t *mp, tmpfs_node_t *tfn, vnode_t **vp) {
...
+  bzero(dirent, sizeof(tmpfs_dirent_t));
+
   dirent->tfd_node = NULL;
   dirent->tfd_namelen = namelen;
   memcpy(dirent->tfd_name, name, namelen);
   dirent->tfd_name[namelen] = 0;
 
   *dep = dirent;
-  return 0;
+  return error;
```
Comment (pj1031999) [inline sys/kern/tmpfs.c:720]: If I understand correctly here we always return 0.
Maybe it will be better to write it explicite?

**Review by pj1031999 — APPROVED** [review]
LGTM

File: sys/kern/tmpfs.c (hunk 1 above)
Comment (xThaid) [reply sys/kern/tmpfs.c]: Makes sense to me.

File: sys/kern/tmpfs.c [hunk 2]
```diff
@@ -48,63 +90,170 @@ typedef struct tmpfs_node {
...
+  return newar;
+}
+
+static mem_arena_t *mem_find_ptr_arena(mem_arenalist_t *al, void *ptr) {
+  mem_arena_t *arena;
+  STAILQ_FOREACH(arena, al, tma_link) {
+    if ((intptr_t)ptr > (intptr_t)arena &&
+        (intptr_t)ptr < (intptr_t)arena + ARENA_SIZE) {
+      return arena;
+    }
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/kern/tmpfs.c]: Good point.

File: sys/kern/tmpfs.c [hunk 3]
```diff
@@ -48,63 +90,170 @@ typedef struct tmpfs_node {
...
+    if ((uintptr_t)ptr > (uintptr_t)arena &&
+        (uintptr_t)ptr < (uintptr_t)arena + ARENA_SIZE) {
+      return arena;
+    }
   }
+  return NULL;
+}
 
-  return 0;
+static blkptr_t mem_arena_alloc_dblk(mem_arena_t *arena) {
```
Comment (j-piecuch) [inline sys/kern/tmpfs.c:177]: Functions that read and/or modify state that is shared across multiple vnodes must be protected by a lock. This state includes the arena list and the arenas themselves.

**Review by j-piecuch — CHANGES_REQUESTED** [review]

File: sys/kern/tmpfs.c [hunk 4]
```diff
@@ -48,63 +90,170 @@ typedef struct tmpfs_node {
...
-    v->tfn_reg.nblocks++;
-    *blkptrp = blkptr;
+  /* bitmaps of free items */
+  bitstr_t tma_inode_bm[bitstr_size(ARENA_INODE_CNT)];
+  bitstr_t tma_dblock_bm[bitstr_size(ARENA_DATA_BLOCKS)];
+
+  /* pointers to first item */
+  tmpfs_node_t *tma_inodes;
+  blkptr_t tma_dblocks;
+} mem_arena_t;
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/kern/tmpfs.c:133]: Make sense to me. Added static assertion.

File: sys/kern/tmpfs.c [hunk 5]
```diff
@@ -48,63 +90,170 @@ typedef struct tmpfs_node {
...
+  int index;
+  assert(arena->tma_ninodes > 0);
+
+  bit_ffs(arena->tma_inode_bm, ARENA_INODE_CNT, &index);
+  bit_clear(arena->tma_inode_bm, index);
+  assert(index != -1);
+  arena->tma_ninodes--;
+
+  tmpfs_node_t *node = arena->tma_inodes + index;
+  try_alloc_frame((vaddr_t)node);
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/kern/tmpfs.c:210]: You're right, it would. Fixed.

File: sys/kern/tmpfs.c [hunk 6]
```diff
@@ -48,63 +90,170 @@ typedef struct tmpfs_node {
...
+  tmpfs_node_t *tma_inodes;
+  blkptr_t tma_dblocks;
+} mem_arena_t;
+
+static void try_alloc_frame(vaddr_t va) {
+  paddr_t pap;
+  va &= ~(PAGESIZE - 1); /* align address to the page size */
+  if (!pmap_extract(pmap_kernel(), va, &pap))
+    kva_map(va, BLOCK_SIZE, M_ZERO);
+}
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/kern/tmpfs.c:140]: Now you convinced me that I don't like neither. Changed to more self-descriptive.

File: sys/kern/tmpfs.c [hunk 7]
```diff
@@ -536,15 +673,44 @@ static int tmpfs_get_vnode(mount_t *mp, tmpfs_node_t *tfn, vnode_t **vp) {
...
+
+  if ((error = tmpfs_resize(tfm, tfn, tfn->tfn_size + BLOCK_SIZE)))
+    return error;
+
+  blkptr_t blk = *tmpfs_get_blk(tfn, BLKNO(tfn->tfn_size) - 1);
+
+  size_t ndirent = BLOCK_SIZE / sizeof(tmpfs_dirent_t);
+  for (size_t i = 0; i < ndirent; i++) {
+    tmpfs_dirent_t *de =
+      (tmpfs_dirent_t *)(((uint8_t *)blk) + i * sizeof(tmpfs_dirent_t));
```
[the remark this answers is not in the document]
Comment (xThaid) [reply sys/kern/tmpfs.c:688]: Good remark. Done.

File: sys/kern/tmpfs.c (hunk 3 above)
Comment (xThaid) [reply sys/kern/tmpfs.c:181]: What an embarrassing oversight on my part. Fixed.

Comment (xThaid) [reply sys/kern/tmpfs.c:181]: It is now protected by the mount point's lock.

**Review by j-piecuch — APPROVED** [review]

**Review by cahirwpz — APPROVED** [review]
After many cosmetic changes LGTM.

**Pull request merged**